format/require.go

```diff
@@ -23,31 +23,131 @@ func (p *Printer) markRequires(n parse.Node) {
+			for _, ref := range []*referList{r.refer, r.referMacros} {
+				for _, n := range ref.origRefer {
+					if n, ok := n.(*parse.SymbolNode); ok {
+						p.refers[n.Val] = r.name
+					}
+				}
+				for ref := range ref.refer {
+					p.refers[ref] = r.name
 				}
-			}
-			for ref := range r.refer {
-				p.refers[ref] = r.name
 			}
 		}
 	}
 }
 
-type require struct {
-	name     string
-	as       map[string]struct{}
-	referAll bool
+// referList represents a list of symbols following either a :refer or
+// :refer-macros keyword.
+type referList struct {
 	// origRefer is the original refer vector/list.
 	// It is preserved and reused if possible, but if two refer lists are
 	// merged, then refer is used instead (and origRefer is set to nil).
 	origRefer []parse.Node
 	refer     map[string]struct{}
+}
+
+func (rl *referList) merge(rl1 *referList) {
+	if rl1 == nil {
+		return
+	}
+	if rl1.refer != nil {
+		panic("merge arg has non-nil refer")
+	}
+	if len(rl1.origRefer) == 0 {
+		return
+	}
+	// If the current list has *no* requires, just copy the origRefer
+	// list from the other.
+	if rl.origRefer == nil && rl.refer == nil {
+		rl.origRefer = rl1.origRefer
+		return
+	}
+	// Otherwise, move everything into rl.refer (if not already moved by
+	// a previous merge) and copy each symbol from the other list.
+	if rl.origRefer != nil {
+		rl.extractOrigRefer()
+	}
+	for _, n := range rl1.origRefer {
+		n, ok := n.(*parse.SymbolNode)
+		if !ok {
+			continue
+		}
+		rl.refer[n.Val] = struct{}{}
+	}
+}
+
+// render returns a parse.VectorNode for all symbols in this referList,
+// or nil if there are no symbols.
+func (rl *referList) render() *parse.VectorNode {
+	if rl.origRefer != nil {
+		return &parse.VectorNode{Nodes: rl.origRefer}
+	}
+	if len(rl.refer) == 0 {
+		return nil
+	}
+	var refs []parse.Node
+	for _, s := range sortStringSet(rl.refer) {
+		refs = append(refs, &parse.SymbolNode{Val: s})
+	}
+	return &parse.VectorNode{Nodes: refs}
+}
+
+// extractOrigRefer moves all symbols in the origRefer slice
+// into rl.refer map.
+func (rl *referList) extractOrigRefer() {
+	rl.refer = make(map[string]struct{})
+	for _, n := range rl.origRefer {
+		if n, ok := n.(*parse.SymbolNode); ok {
+			rl.refer[n.Val] = struct{}{}
+		}
+	}
+	rl.origRefer = nil
+}
+
+// removeUnused removes all symbols from this referList that aren't
+// present in sc.
+func (rl *referList) removeUnused(sc *symbolCache) {
+	if rl.origRefer != nil {
+		// If origRefer doesn't have any unused elements, leave it
+		// alone. Otherwise, rewrite it as a refer and handle below.
+		for _, n := range rl.origRefer {
+			n, ok := n.(*parse.SymbolNode)
+			if !ok {
+				continue
+			}
+			if !sc.usesSym(n.Val) {
+				rl.extractOrigRefer()
+				break
+			}
+		}
+	}
+	for ref := range rl.refer {
+		if !sc.usesSym(ref) {
+			delete(rl.refer, ref)
+		}
+	}
+}
+
+type require struct {
+	name     string
+	as       map[string]struct{}
+	referAll bool
+
+	refer       *referList
```

Comment:
Very minor/optional, but you don't really need the pointers here -- you could do

    refer       referList
    referMacros referList

and then you don't need to separately allocate these. You'd use `&r.refer` in a few places.